share/p2p/shrexnd/server.go

```diff
@@ -95,90 +141,65 @@ func (srv *Server) handleNamespacedData(ctx context.Context, stream network.Stre
 	var req pb.GetSharesByNamespaceRequest
 	_, err = serde.Read(stream, &req)
 	if err != nil {
-		logger.Warnw("server: reading request", "err", err)
-		stream.Reset() //nolint:errcheck
-		return
+		return nil, fmt.Errorf("reading request: %w", err)
+
 	}
-	logger = logger.With("namespace", hex.EncodeToString(req.Namespace), "hash", share.DataHash(req.RootHash).String())
-	logger.Debugw("server: new request")
 
+	logger.Debugw("server: new request")
 	err = stream.CloseRead()
 	if err != nil {
 		logger.Debugw("server: closing read side of the stream", "err", err)
 	}
 
 	err = validateRequest(req)
 	if err != nil {
-		logger.Warnw("server: invalid request", "err", err)
-		stream.Reset() //nolint:errcheck
-		return
+		return nil, fmt.Errorf("invalid request: %w", err)
 	}
+	return &req, nil
+}
 
-	ctx, cancel := context.WithTimeout(ctx, srv.params.HandleRequestTimeout)
-	defer cancel()
-
-	dah, err := srv.store.GetDAH(ctx, req.RootHash)
+func (srv *Server) getNamespaceData(ctx context.Context,
+	hash share.DataHash, namespace share.Namespace) (share.NamespacedShares, pb.StatusCode, error) {
+	dah, err := srv.store.GetDAH(ctx, hash)
 	if err != nil {
 		if errors.Is(err, eds.ErrNotFound) {
-			logger.Warn("server: DAH not found")
-			srv.respondNotFoundError(ctx, logger, stream)
-			return
+			return nil, pb.StatusCode_NOT_FOUND, nil
 		}
-		logger.Errorw("server: retrieving DAH", "err", err)
-		srv.respondInternalError(ctx, logger, stream)
-		return
+		return nil, pb.StatusCode_INTERNAL, fmt.Errorf("retrieving DAH: %w", err)
 	}
 
-	shares, err := srv.getter.GetSharesByNamespace(ctx, dah, req.Namespace)
-	switch {
-	case errors.Is(err, share.ErrNotFound):
-		logger.Warn("server: nd not found")
-		srv.respondNotFoundError(ctx, logger, stream)
-		return
-	case err != nil:
-		logger.Errorw("server: retrieving shares", "err", err)
-		srv.respondInternalError(ctx, logger, stream)
-		return
-	}
-
-	resp := namespacedSharesToResponse(shares)
-	srv.respond(ctx, logger, stream, resp)
-}
-
-// validateRequest checks correctness of the request
-func validateRequest(req pb.GetSharesByNamespaceRequest) error {
-	if err := share.Namespace(req.Namespace).ValidateForData(); err != nil {
-		return err
+	shares, err := srv.getter.GetSharesByNamespace(ctx, dah, namespace)
+	if err != nil {
+		return nil, pb.StatusCode_INTERNAL, fmt.Errorf("retrieving shares: %w", err)
 	}
-	if len(req.RootHash) != sha256.Size {
-		return fmt.Errorf("incorrect root hash length: %v", len(req.RootHash))
+
+	if shares.IsEmpty() {
+		return shares, pb.StatusCode_NAMESPACE_NOT_FOUND, nil
```

Comment:
Do we need still need StatusCode_NAMESPACE_NOT_FOUND if not found is a success case now?

Reply:
I would still consider the case where the namespace is out of range as an invalid request, though, as the client can figure that itself.

Reply:
True, StatusCode_NAMESPACE_NOT_FOUND is no longer needed. It is removed from code now. 

To allow handling invalid requests the way you suggested, it is required to add `filterRootsByNamespace` to shrex-nd client. Currently it is not easy due to code structure and dependancy graph and will require some reorganisation of share,getter,ipld packages. I think it will be too much for this PR, so for now it will be always `Ok` from server if namespace is not found. Not as clean as suggestion, but should be fine